js/ui/calendar/ui.calendar.js

```diff
@@ -241,61 +249,128 @@ const Calendar = Editor.inherit({
         this.option(optionName, dateSerialization.serializeDate(optionValue, serializationFormat));
     },
 
-    _moveCurrentDate: function(offset, baseDate) {
-        let currentDate = baseDate || new Date(this.option('currentDate'));
+    _shiftDate: function(zoomLevel, date, offset, reverse) {
+        switch(zoomLevel) {
+            case ZOOM_LEVEL.MONTH:
+                date.setDate(date.getDate() + offset * reverse);
+                break;
+            case ZOOM_LEVEL.YEAR:
+                date.setMonth(date.getMonth() + offset * reverse);
+                break;
+            case ZOOM_LEVEL.DECADE:
+                date.setFullYear(date.getFullYear() + offset * reverse);
+                break;
+            case ZOOM_LEVEL.CENTURY:
+                date.setFullYear(date.getFullYear() + 10 * offset * reverse);
+                break;
+        }
+    },
+
+    _moveCurrentDateByOffset: function(offset) {
+        const baseDate = this.option('currentDate');
+        let currentDate = new Date(baseDate);
+        const zoomLevel = this.option('zoomLevel');
+        this._shiftDate(zoomLevel, currentDate, offset, 1);
+
         const maxDate = this._getMaxDate();
         const minDate = this._getMinDate();
+
+        let isDateForwardInNeighborView = this._areDatesInNeighborView(zoomLevel, currentDate, baseDate);
+        let isDateForwardInRange = inRange(currentDate, minDate, maxDate) && isDateForwardInNeighborView;
+        const dateForward = new Date(currentDate);
+
+        while(isDateForwardInRange) {
+            if(isDateForwardInRange && !this._view.isDateDisabled(dateForward)) {
+                currentDate = dateForward;
+                break;
+            }
+
+            this._shiftDate(zoomLevel, dateForward, offset, 1);
+
+            isDateForwardInNeighborView = this._areDatesInNeighborView(zoomLevel, dateForward, baseDate);
+            isDateForwardInRange = inRange(dateForward, minDate, maxDate) && isDateForwardInNeighborView;
+        }
+
+        if(this._view.isDateDisabled(baseDate) || this._view.isDateDisabled(currentDate)) {
+            this._waitRenderView(offset > 0 ? 1 : -1);
+        } else {
+            this.option('currentDate', currentDate);
+        }
+    },
+
+    _areDatesInSameView(zoomLevel, date1, date2) {
+        switch(zoomLevel) {
+            case ZOOM_LEVEL.MONTH:
+                return date1.getMonth() === date2.getMonth();
+            case ZOOM_LEVEL.YEAR:
+                return date1.getYear() === date2.getYear();
+            case ZOOM_LEVEL.DECADE:
+                return parseInt(date1.getYear() / 10) === parseInt(date2.getYear() / 10);
+            case ZOOM_LEVEL.CENTURY:
+                return parseInt(date1.getYear() / 100) === parseInt(date2.getYear() / 100);
+        }
+    },
+
+    _areDatesInNeighborView(zoomLevel, date1, date2) {
+        const monthMinDistance = (a, b) => {
```

Comment:
To use arrow functions you need to refactor this file to ES6 and replace `require` with `import`
This code will break in IE11.

Reply:
Thanks, I have replaced require to import